CommonTools/RecoUtils/plugins/LeptonFSRProducer.cc

```diff
@@ -0,0 +1,284 @@
+        }
+        if (photon_relIso03 > isoCut) {
+          break;  // break loop on electrons -> photon will be skipped
+        }
+        // Check that photon is not in footprint of an electron
+        pat::PackedCandidateRef pfcandRef = pat::PackedCandidateRef(pfcands, pc - pfcands->begin());
+        if (electronFootprintVeto(pfcandRef, electronsForVeto)) {
+          break;  // break loop on electrons -> photon will be skipped
+        }
+
+        // Candidate matching
+        dRMin = dR;
+        closestEle = eleIdx;
+        closestMu = -1;  // reset match to muons
+      }
+    }  // end loop on electrons
+
+    if (closestMu >= 0 || closestEle >= 0) {
+      // Add FSR photon to the output collection
+      double dRET2 = dRMin / pc->pt() / pc->pt();
+      int iPhoton = fsrPhotons->size();
+      fsrPhotons->push_back(pat::GenericParticle(*pc));
+      fsrPhotons->back().addUserFloat("relIso03", photon_relIso03);
+      fsrPhotons->back().addUserFloat("dROverEt2", dRET2);
+
+      if (closestMu >= 0) {
+        fsrPhotons->back().addUserCand("associatedMuon", reco::CandidatePtr(muons, closestMu));
+        // Store the backlink to the photon: choose the lowest-dRET2 photon for each mu...
+        if (dRET2 < muPhotonDRET2[closestMu]) {
+          muPhotonDRET2[closestMu] = dRET2;
+          muPhotonIdxs[closestMu] = iPhoton;
+        }
+      } else if (closestEle >= 0) {
+        // ...and same for eles
+        fsrPhotons->back().addUserCand("associatedElectron", reco::CandidatePtr(electrons, closestEle));
+        if (dRET2 < elePhotonDRET2[closestEle]) {
+          elePhotonDRET2[closestEle] = dRET2;
+          elePhotonIdxs[closestEle] = iPhoton;
+        }
+      }
+    }
+  }  // end of loop over pfCands
+
+  edm::OrphanHandle<std::vector<pat::GenericParticle>> oh = iEvent.put(std::move(fsrPhotons));
+
+  {
+    std::unique_ptr<edm::ValueMap<int>> bareIdx(new edm::ValueMap<int>());
+    edm::ValueMap<int>::Filler fillerBareIdx(*bareIdx);
+    fillerBareIdx.insert(muons, muPhotonIdxs.begin(), muPhotonIdxs.end());
+    fillerBareIdx.fill();
+    iEvent.put(std::move(bareIdx), "muFsrIndex");
+  }
+
+  {
+    std::unique_ptr<edm::ValueMap<int>> bareIdx(new edm::ValueMap<int>());
+    edm::ValueMap<int>::Filler fillerBareIdx(*bareIdx);
+    fillerBareIdx.insert(electrons, elePhotonIdxs.begin(), elePhotonIdxs.end());
+    fillerBareIdx.fill();
+    iEvent.put(std::move(bareIdx), "eleFsrIndex");
+  }
+}
+
+double LeptonFSRProducer::computeRelativeIsolation(const pat::PackedCandidate& photon,
+                                                   const pat::PackedCandidateCollection& pfcands,
+                                                   const double& isoConeMax,
+                                                   const double& isoConeMin) const {
+  double ptsum = 0;
+
+  for (const auto& pfcand : pfcands) {
+    // Isolation cone
+    double dRIsoCone = deltaR(photon.eta(), photon.phi(), pfcand.eta(), pfcand.phi());
```

Comment:
`deltaR2` would be computationally cheaper

Reply:
Indeed but that would make code less readable, and the same comment applies to the class we are replacing (MuonFSRProducer/MuonFSRAssociator). We can do that if you insist but I have a preference in keeping it more readable.

Reply:
Uhm... I did not make the same comment in the other place where `deltaR` is computed exactly because it could have been less readable there, but here `dRIsoCone` is only used in the comparisons: if you pass (e.g.) `isoConeMax2` and `isoConeMin2` (and modify accordingly L250) it would be equally readable, and one square root per candidate and per lepton can be avoided

Reply:
OK, done